ddclient

```diff
@@ -2298,57 +2281,72 @@ sub get_ip {
+                         (?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))\b/xai;  ## last one
+    return ($1 // '') =~ s/\b0+\B//gr;                                     ## remove leading zeros
 }
 
 ######################################################################
-## is_ipv6() validates if string is valid IPv6 address
+## is_ipv6() validates if string is valid IPv6 address with no preceding
+## or trailing spaces/characters and no embedded leading zeros.
 ######################################################################
 sub is_ipv6 {
     my ($value) = @_;
-    # This little gem from http://home.deds.nl/~aeron/regex/
-    return $value =~ /^(((?=.*(::))(?!.*\3.+\3))\3?|([\dA-F]{1,4}(\3|:\b|$)|\2))(?4){5}((?4){2}|(((2[0-4]|1\d|[1-9])?\d|25[0-5])\.?\b){4})\z/ai;
+    return (length($value // '') != 0) && ($value eq (ipv6_match($value) // ''));
 }
 
 ######################################################################
-## ipv6_match determine ipv6 address from given string and return them
+## ipv6_match() extracts the first valid IPv6 address from given string
+## IPv6 must be in standard or compressed format.
+## Mixed IPv6/IPv4 not supported.
+## Accepts leading zeros in the address but removes them in returned value
 ######################################################################
 sub ipv6_match {
-    my $content = shift;
-    my $omits;
-    my $ip          = "";
-    my $linenumbers = 0;
-
-    my @values = split('\n', $content);
-    foreach my $val (@values) {
-        next unless $val =~ /((:{0,2}[A-F0-9]{1,4}){0,7}:{1,2}[A-F0-9]{1,4})/ai; # invalid char
-        my $parsed = $1;
-
-        # check for at least 7 colons
-        my $count_colon = () = $parsed =~ /:/g;
-        if ($count_colon != 7) {
-            # or one double colon
-            my $count_double_colon = () = $parsed =~ /::/g;
-            if ($count_double_colon != 1) {
-                next
-            }
-        }
-        return $parsed;
-    }
-    return;
+    (shift // '') =~ /(?<![:.\w])                                     ## Negative Look behind for word boundry
+                      ( (?:[0-9A-F]{1,4}:){7}[0-9A-F]{1,4}            ## Standard format
+                       |(?=(?:[0-9A-F]{0,4}:){0,7}[0-9A-F]{0,4}       ## OR compressed with at most 7 colons
+                        (?![:.\w]))                                   ##   and negative lookahead for boundry
+                        (?:(?:[0-9A-F]{1,4}:){1,7}|:)(?:(?::[0-9A-F]{1,4}){1,7}|:)  ## and only 1 double-colon
+                       |(?:[0-9A-F]{1,4}:){7}:                        ## OR compressed with 8 colons (double at end)
+                       |:(?::[0-9A-F]{1,4}){7}                        ## OR compressed with 8 colons (double at front)
+                      )                                               ## End of first capture group
+                      (?![:.\w])/xai;                                 ## Negative lookahead for boundry
```

Comment:
This is complicated enough that I don't feel comfortable merging this without unit tests.

Reply:
How would you like to do that?  There are various places where I have found lists of valid and invalid IPv6 addresses which can be used for test purposes.  The lists are not very good however -- they do test for valid formatting (number of colons, colons in wrong places, etc) but don't include examples of GUA, ULA and LLA addresses which I think a good test bucket should include, certainly for our purposes.

Reply:
> How would you like to do that?

You'll have to wait for PR #161 to be merged.  Until then, split this into two PRs: one that doesn't require tests and one that does.